Support generic inference through function pointer types

333fred opened the pull request.

333fred commented:
@AlekseyTs @dotnet/roslyn-compiler for review.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -556,7 +556,7 @@ private void InferTypeArgsFirstPhase(Binder binder, ref HashSet<DiagnosticInfo>
                 BoundExpression argument = _arguments[arg];
                 TypeWithAnnotations target = _formalParameterTypes[arg];
-                ExactOrBoundsKind kind = GetRefKind(arg).IsManagedReference() || target.Type.IsPointerType() ? ExactOrBoundsKind.Exact : ExactOrBoundsKind.LowerBound;
+                ExactOrBoundsKind kind = GetRefKind(arg).IsManagedReference() || target.Type.IsPointerOrFunctionPointer() ? ExactOrBoundsKind.Exact : ExactOrBoundsKind.LowerBound;
```
>IsPointerOrFunctionPointer [](start = 93, length = 26)

I would expect that we would support inference somewhat similar to delegates

333fred replied:
Added a note to [https://github.com/dotnet/roslyn/issues/39865](https://github.com/dotnet/roslyn/issues/39865).

---
In reply to: [406388004](https://github.com/dotnet/roslyn/pull/43041#discussion_r406388004) [](ancestors = 406388004)

AlekseyTs commented:
What inference are we going to support for array initializers and conditional expressions? Perhaps we should start with that and move to more complex type argument inference.

AlekseyTs commented:
>     // PROTOTYPE(func-ptr): Support generic substitution and retargeting

Perhaps we should implement this first?

---
Refers to: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:17 in 123406e. [](commit_id = 123406e0b12838969cb963a6dd853dc30a7fd3b9, deletion_comment = False)

333fred commented:
>     // PROTOTYPE(func-ptr): Support generic substitution and retargeting

I think this PR does actually address this comment, I just forgot to remove it.

---
In reply to: [611681794](https://github.com/dotnet/roslyn/pull/43041#issuecomment-611681794) [](ancestors = 611681794)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:17 in 123406e. [](commit_id = 123406e0b12838969cb963a6dd853dc30a7fd3b9, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs:
```diff
@@ -232,6 +235,31 @@ private PointerTypeSymbol SubstitutePointerType(PointerTypeSymbol t)
+        {
+            bool paramTypesChanged = false;
+            var substitutedParamTypes = ArrayBuilder<TypeWithAnnotations>.GetInstance(f.Signature.ParameterCount);
+            foreach (var param in f.Signature.Parameters)
```
>foreach (var param in f.Signature.Parameters) [](start = 12, length = 45)

```SubstituteTypes(ParameterTypesWithAnnotations)```?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs:
```diff
@@ -232,6 +235,31 @@ private PointerTypeSymbol SubstitutePointerType(PointerTypeSymbol t)
+
+            if (paramTypesChanged || !f.Signature.ReturnTypeWithAnnotations.IsSameAs(substitutedReturnType))
+            {
+                f = f.SubstituteTypeSymbol(substitutedReturnType, substitutedParamTypes);
```
>substitutedParamTypes [](start = 66, length = 21)

It is probably better to pass an immutable array as we do for substituted generic type arguments

gafter commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs:
```diff
@@ -556,7 +556,7 @@ private void InferTypeArgsFirstPhase(Binder binder, ref HashSet<DiagnosticInfo>
                 BoundExpression argument = _arguments[arg];
                 TypeWithAnnotations target = _formalParameterTypes[arg];
-                ExactOrBoundsKind kind = GetRefKind(arg).IsManagedReference() || target.Type.IsPointerType() ? ExactOrBoundsKind.Exact : ExactOrBoundsKind.LowerBound;
+                ExactOrBoundsKind kind = GetRefKind(arg).IsManagedReference() || target.Type.IsPointerOrFunctionPointer() ? ExactOrBoundsKind.Exact : ExactOrBoundsKind.LowerBound;
```
>IsManagedReference [](start = 57, length = 18)

I suggest adding a PROTOTYPE comment here that the behavior for function pointers will have to be aligned with a specification once one is written.

333fred replied:
Added a comment here: [https://github.com/dotnet/roslyn/issues/39865](https://github.com/dotnet/roslyn/issues/39865)

---
In reply to: [406414229](https://github.com/dotnet/roslyn/pull/43041#discussion_r406414229) [](ancestors = 406414229)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -726,5 +727,91 @@ void M(delegate*<delegate*<object, void>, void> param1, delegate*<delegate*<obje
+
+            Assert.NotSame(methodSymbol, methodSymbol.OriginalDefinition);
+            Assert.Equal(SpecialType.System_String, methodSymbol.TypeArguments[0].SpecialType);
+            var functionPointer = (FunctionPointerTypeSymbol)methodSymbol.Parameters[0].Type;
```
>(FunctionPointerTypeSymbol) [](start = 34, length = 27)

This feels wrong. How could this cast succeed on an ITypeSymbol?

333fred replied:
This is a symbol from the `PublicModel` namespace. I can qualify it to make it obvious.

---
In reply to: [406417601](https://github.com/dotnet/roslyn/pull/43041#discussion_r406417601) [](ancestors = 406417601)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -726,5 +727,91 @@ void M(delegate*<delegate*<object, void>, void> param1, delegate*<delegate*<obje
+
+            Assert.NotSame(methodSymbol, methodSymbol.OriginalDefinition);
+            Assert.Equal(SpecialType.System_String, methodSymbol.TypeArguments[0].SpecialType);
+            var functionPointer = (FunctionPointerTypeSymbol)methodSymbol.Parameters[0].Type;
```
>(FunctionPointerTypeSymbol) [](start = 34, length = 27)

Same here

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -726,5 +727,91 @@ void M(delegate*<delegate*<object, void>, void> param1, delegate*<delegate*<obje
+
+            Assert.NotSame(methodSymbol, methodSymbol.OriginalDefinition);
+            Assert.Equal(TypeKind.TypeParameter, methodSymbol.TypeArguments[0].TypeKind);
+            var functionPointer = (FunctionPointerTypeSymbol)methodSymbol.Parameters[0].Type;
```
>(FunctionPointerTypeSymbol) [](start = 34, length = 27)

Same here.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -726,5 +727,91 @@ void M(delegate*<delegate*<object, void>, void> param1, delegate*<delegate*<obje
+            Assert.NotSame(methodSymbol, methodSymbol.OriginalDefinition);
+            Assert.Equal(TypeKind.TypeParameter, methodSymbol.TypeArguments[0].TypeKind);
+            var functionPointer = (FunctionPointerTypeSymbol)methodSymbol.Parameters[0].Type;
+            Assert.Equal(TypeKind.TypeParameter, functionPointer.Signature.ReturnType.TypeKind);
```
>Assert.Equal(TypeKind.TypeParameter, functionPointer.Signature.ReturnType.TypeKind); [](start = 12, length = 84)

Perhaps we should assert more than that. What type parameter is this?

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -726,5 +727,91 @@ void M(delegate*<delegate*<object, void>, void> param1, delegate*<delegate*<obje
         }
+
+        [Fact]
+        public void FunctionPointerParameterTypeInference()
```
>FunctionPointerParameterTypeInference [](start = 20, length = 37)

Should also test an error case, for example between ```delegate*<string, void>``` and ```delegate*<object, void>```

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -726,5 +727,91 @@ void M(delegate*<delegate*<object, void>, void> param1, delegate*<delegate*<obje
+        }
+
+        [Fact]
+        public void FunctionPointerReturnTypeInference()
```
>FunctionPointerReturnTypeInference [](start = 20, length = 34)

Similar, should test an error scenario

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -726,5 +727,91 @@ void M(delegate*<delegate*<object, void>, void> param1, delegate*<delegate*<obje
+            Assert.Equal(TypeKind.TypeParameter, functionPointer.Signature.ReturnType.TypeKind);
+            Assert.Equal(TypeKind.TypeParameter, functionPointer.Signature.Parameters[0].Type.TypeKind);
+        }
     }
```
>} [](start = 4, length = 1)

I think we should also test a scenarios when we infer a function pointer for a type parameter. Including scenarios when there are more than one candidate and the types are:
- identical
- convertible
- not convertible
- with just calling convention differences
- etc.

333fred replied:
> I think we should also test a scenarios when we infer a function pointer for a type parameter. 

Function pointers cannot be type parameters.

---
In reply to: [406422979](https://github.com/dotnet/roslyn/pull/43041#discussion_r406422979) [](ancestors = 406422979)

AlekseyTs replied:
> Function pointers cannot be type parameters.

But we are going to try to infer something. Let's see what is going to happen and what do we infer.
Also, consider adding tests for inference for array initializers and for conditional expressions, it might just work.

---
In reply to: [406455969](https://github.com/dotnet/roslyn/pull/43041#discussion_r406455969) [](ancestors = 406455969,406422979)

AlekseyTs replied:
Also, we might avoid an error if we infer directly form an array of function pointers rather than from a function pointer. Might want to test scenarios like that as well.

---
In reply to: [406467051](https://github.com/dotnet/roslyn/pull/43041#discussion_r406467051) [](ancestors = 406467051,406455969,406422979)

AlekseyTs replied:
It doesn't looks like there are test covering:
- not convertible (for arrays)
- with just calling convention differences (for arrays and pointers)

---
In reply to: [406468735](https://github.com/dotnet/roslyn/pull/43041#discussion_r406468735) [](ancestors = 406468735,406467051,406455969,406422979)

333fred replied:
Just calling convention differences are covered by `FunctionPointerAsTypeArgument` for pointer types.

---
In reply to: [406881794](https://github.com/dotnet/roslyn/pull/43041#discussion_r406881794) [](ancestors = 406881794,406468735,406467051,406455969,406422979)

AlekseyTs replied:
> Just calling convention differences are covered by FunctionPointerAsTypeArgument for pointer types.

Please add coverage for arrays too

---
In reply to: [406910516](https://github.com/dotnet/roslyn/pull/43041#discussion_r406910516) [](ancestors = 406910516,406881794,406468735,406467051,406455969,406422979)

AlekseyTs replied:
This is marked as resolved, but I still don't see the tests for method type inference
- not convertible (for arrays)
- with just calling convention differences (for arrays)

---
In reply to: [406933066](https://github.com/dotnet/roslyn/pull/43041#discussion_r406933066) [](ancestors = 406933066,406910516,406881794,406468735,406467051,406455969,406422979)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -6,6 +6,7 @@
 #nullable enable
 
 using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Symbols.PublicModel;
```
Is this needed?

AlekseyTs commented:
Done with review pass (iteration 1)

333fred commented:
Adding tests.

---
In reply to: [611678140](https://github.com/dotnet/roslyn/pull/43041#issuecomment-611678140) [](ancestors = 611678140)

333fred commented:
@AlekseyTs @gafter addressed feedback.

---
In reply to: [611708057](https://github.com/dotnet/roslyn/pull/43041#issuecomment-611708057) [](ancestors = 611708057)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs:
```diff
@@ -232,6 +235,19 @@ private PointerTypeSymbol SubstitutePointerType(PointerTypeSymbol t)
+            ImmutableArray<TypeWithAnnotations> substitutedParamTypes = SubstituteTypes(f.Signature.ParameterTypesWithAnnotations);
+            var substitutedReturnType = f.Signature.ReturnTypeWithAnnotations.SubstituteType(this);
+
+            if (substitutedParamTypes != f.Signature.ParameterTypesWithAnnotations || !f.Signature.ReturnTypeWithAnnotations.IsSameAs(substitutedReturnType))
```
>f.Signature.ParameterTypesWithAnnotations  [](start = 41, length = 42)

This feels fragile. We are taking dependency that ```ParameterTypesWithAnnotations``` returns the same array the second time, however this isn't API's contract, but rather an implementation detail, an optimization. Consider capturing result of first access in a local and avoiding second access.

AlekseyTs commented:
>     // PROTOTYPE(func-ptr): Support generic substitution and retargeting

This PR doesn't implement retargeting, should keep relevant portion of the comment.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:17 in e1861c9. [](commit_id = e1861c9f5e95f0cc9d085e7a8a54af278dfdaa38, deletion_comment = True)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:
```diff
@@ -130,14 +132,25 @@ internal override bool GetUnificationUseSiteDiagnosticRecursive(ref DiagnosticIn
         {
-            // PROTOTYPE(func-ptr): Implement
-            throw new NotImplementedException();
+            Debug.Assert(this.Equals(other, TypeCompareKind.IgnoreDynamicAndTupleNames | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes));
```
>TypeCompareKind.IgnoreDynamicAndTupleNames | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes [](start = 44, length = 101)

I think ```AllIgnoreOptions``` is correct option to use here and is more robust long term.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -10,6 +10,7 @@
 using System.Linq;
 using System.Reflection;
 using Microsoft.Cci;
+using Microsoft.CodeAnalysis;
```
Why do we need this? We are inside this namespace.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +96,120 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                        originalParam.RefKind,
+                        originalParam.Ordinal,
+                        containingSymbol: this,
+                        originalParam.RefCustomModifiers));
```
>originalParam.RefCustomModifiers [](start = 24, length = 32)

I think these are subject for substitution as well?

333fred replied:
Comment added to tracking issue.

---
In reply to: [406851037](https://github.com/dotnet/roslyn/pull/43041#discussion_r406851037) [](ancestors = 406851037)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +96,120 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                this.CallingConvention,
+                this.RefKind,
+                substitutedReturnType,
+                this.RefCustomModifiers,
```
>this.RefCustomModifiers [](start = 16, length = 23)

I think these are subject for substitution as well. Please add a test.

333fred replied:
I have no idea what could cause these to be substituted. Can you please show a scenario?

---
In reply to: [406852356](https://github.com/dotnet/roslyn/pull/43041#discussion_r406852356) [](ancestors = 406852356)

AlekseyTs replied:
> I have no idea what could cause these to be substituted. Can you please show a scenario?

I guess, the modifier is a generic type closed over ```T```. When ```T``` is substituted, the modifier should change accordingly.

---
In reply to: [406863326](https://github.com/dotnet/roslyn/pull/43041#discussion_r406863326) [](ancestors = 406863326,406852356)

333fred replied:
Adding a test is tracked by the tracking issue.

---
In reply to: [406865984](https://github.com/dotnet/roslyn/pull/43041#discussion_r406865984) [](ancestors = 406865984,406863326,406852356)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +96,120 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                var builder = ArrayBuilder<TypeWithAnnotations>.GetInstance(_parameters.Length);
+                var paramVariance = variance switch
+                {
+                    VarianceKind.In => VarianceKind.Out,
```
>VarianceKind.In => VarianceKind.Out [](start = 20, length = 35)

Does parameter ref-ness matter here?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +96,120 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                    }
+                }
+
+                mergedParameterTypes = builder.ToImmutableAndFree();
```
>builder.ToImmutableAndFree() [](start = 39, length = 28)

Perhaps we should free the builder instead when there are no changes. We also often do not even allocate the builder until there is a change, but that is optional, I guess.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +96,120 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                mergedParameterTypes = builder.ToImmutableAndFree();
+            }
+
+            var mergedReturnType = ReturnTypeWithAnnotations.MergeEquivalentTypes(signature.ReturnTypeWithAnnotations, variance);
```
>var mergedReturnType = ReturnTypeWithAnnotations.MergeEquivalentTypes(signature.ReturnTypeWithAnnotations, variance); [](start = 12, length = 117)

Does return ref-ness matter here?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +96,120 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                    }
+                }
+
+                transformedParameterTypes = builder.ToImmutableAndFree();
```
>builder.ToImmutableAndFree(); [](start = 44, length = 29)

Same comment about avoiding allocating new array when there are no changes.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +96,120 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+            }
+        }
+
+        public FunctionPointerMethodSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnotations, TypeWithAnnotations> transform)
```
>SetNullabilityForReferenceTypes [](start = 43, length = 31)

I assume somewhere we have a function that performs the opposite operation, i.e. builds the transform flags. Could you point me to it?

333fred replied:
It's not currently implemented, see `FunctionPointerTypeSymbol.AddNullableTransforms` (line 111 in this PR)

---
In reply to: [406857994](https://github.com/dotnet/roslyn/pull/43041#discussion_r406857994) [](ancestors = 406857994)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +96,120 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                transformedParameterTypes = builder.ToImmutableAndFree();
+            }
+
+            var transformedReturn = transform(ReturnTypeWithAnnotations);
```
>var transformedReturn = transform(ReturnTypeWithAnnotations); [](start = 12, length = 61)

I think this violates our general approach that transforms are applied to types in the order in which types appear in metadata signature. I believe, in metadata signature return type comes before parameter types.

AlekseyTs replied:
Please add targeted tests nullable encoding/decoding for function pointers in metadata.

---
In reply to: [406858800](https://github.com/dotnet/roslyn/pull/43041#discussion_r406858800) [](ancestors = 406858800)

333fred replied:
Nullable tests are already covered by a prototype comment and are out of scope for this PR. Will add this to the list to test when that is addressed.

---
In reply to: [406859222](https://github.com/dotnet/roslyn/pull/43041#discussion_r406859222) [](ancestors = 406859222,406858800)

AlekseyTs replied:
It is fine to postpone testing, but implementation should be fixed

---
In reply to: [406862984](https://github.com/dotnet/roslyn/pull/43041#discussion_r406862984) [](ancestors = 406862984,406859222,406858800)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3691,6 +3691,108 @@ .maxstack  4
         }
 
+        [Fact]
+        public void InferredArrayInitializer()
```
>InferredArrayInitializer [](start = 20, length = 24)

I think we should also test when:
- there is a difference in return types
- inference fails
- the only difference is calling convention
- etc.

AlekseyTs replied:
This is marked as resolved, but it doesn't look like any tests were added for:
- inference fails
- the only difference is calling convention


---
In reply to: [406862319](https://github.com/dotnet/roslyn/pull/43041#discussion_r406862319) [](ancestors = 406862319)

333fred replied:
See `NoBestTypeArrayInitializer` in `FunctionPointerTests.cs`

---
In reply to: [406969760](https://github.com/dotnet/roslyn/pull/43041#discussion_r406969760) [](ancestors = 406969760,406862319)

AlekseyTs replied:
> See NoBestTypeArrayInitializer in FunctionPointerTests.cs

I am not sure why different scenarios for the same thing are tested in different files. I personally don't see any advantage in splitting feature testing across different assemblies, unless there is a dependency on specific helpers/infrastructure

---
In reply to: [406970490](https://github.com/dotnet/roslyn/pull/43041#discussion_r406970490) [](ancestors = 406970490,406969760,406862319)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3691,6 +3691,108 @@ .maxstack  4
+    {
+        delegate*<object, void> ptr1 = &Print;
+        delegate*<string, void> ptr2 = &Print;
+        delegate*<string, void> ptr3 = b ? ptr1 : ptr2;
```
>delegate*<string, void> ptr3 [](start = 8, length = 28)

Shouldn't this be ```var```? Or perhaps just do  ```(b ? ptr1 : ptr2)(""1"")```?

333fred replied:
`?:` is not target-typed today, and I wanted to show what the actual type that is deduced will be.

---
In reply to: [406863504](https://github.com/dotnet/roslyn/pull/43041#discussion_r406863504) [](ancestors = 406863504)

AlekseyTs replied:
> ?: is not target-typed today, and I wanted to show what the actual type that is deduced will be.

This almost defeats the purpose of the test. I think the signature used by ```calli``` is clear enough.

---
In reply to: [406865958](https://github.com/dotnet/roslyn/pull/43041#discussion_r406865958) [](ancestors = 406865958,406863504)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3691,6 +3691,108 @@ .maxstack  4
+        }
+
+        [Fact]
+        public void BestTypeForConditional()
```
>BestTypeForConditional [](start = 20, length = 22)

Similar suggestions for tests as above.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3691,6 +3691,108 @@ .maxstack  4
+        ptr3(""1"");
+    }
+    static void Main() => M(true);
+}", expectedOutput: "1Object");
```
>1Object [](start = 21, length = 7)

Consider covering other branch too.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -726,5 +729,382 @@ void M(delegate*<delegate*<object, void>, void> param1, delegate*<delegate*<obje
+            Assert.Equal(TypeKind.TypeParameter, functionPointer.Signature.Parameters[0].Type.TypeKind);
+
+            var m2DeclarationSyntax = tree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Skip(1).Single();
+            var declaredSymbol = (IMethodSymbol)model.GetDeclaredSymbol(m2DeclarationSyntax)!;
```
>var declaredSymbol = (IMethodSymbol)model.GetDeclaredSymbol(m2DeclarationSyntax)!; [](start = 12, length = 82)

This looks more complicated than it should be, we can locate the symbol from symbol table, there are a lot of helpers for that.

AlekseyTs replied:
>declaredSymbol [](start = 16, length = 14)

Please assert what method is this to eliminate all the guess work

AlekseyTs commented:
Done with review pass (iteration 3)

333fred commented:
>     // PROTOTYPE(func-ptr): Support generic substitution and retargeting

Added a note to the tracking issue.

---
In reply to: [612113693](https://github.com/dotnet/roslyn/pull/43041#issuecomment-612113693) [](ancestors = 612113693)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:17 in e1861c9. [](commit_id = e1861c9f5e95f0cc9d085e7a8a54af278dfdaa38, deletion_comment = True)

AlekseyTs commented:
>     // PROTOTYPE(func-ptr): Support generic substitution and retargeting

> Added a note to the tracking issue.

This should be addressed before we can merge to master.

---
In reply to: [612184511](https://github.com/dotnet/roslyn/pull/43041#issuecomment-612184511) [](ancestors = 612184511,612113693)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:17 in e1861c9. [](commit_id = e1861c9f5e95f0cc9d085e7a8a54af278dfdaa38, deletion_comment = True)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [612157757](https://github.com/dotnet/roslyn/pull/43041#issuecomment-612157757) [](ancestors = 612157757)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +95,244 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+            Debug.Assert(RefKind == signature.RefKind);
+            var returnVariance = RefKind == RefKind.None ? variance : VarianceKind.None;
+            var mergedReturnType = ReturnTypeWithAnnotations.MergeEquivalentTypes(signature.ReturnTypeWithAnnotations, returnVariance);
+            var mergedReturnRefCustomMods = mergeCustomModifiers(RefCustomModifiers, signature.RefCustomModifiers, returnVariance);
```
> var mergedReturnRefCustomMods = mergeCustomModifiers(RefCustomModifiers, signature.RefCustomModifiers, returnVariance); [](start = 11, length = 120)

This doesn't feel right. Custom modifiers are not part of the language and I cannot even think what would it mean to merge them, we don't even make sure they are compatible, etc. Unless we already have an example of doing this, for example during previous call, we shouldn't be doing this here. Follow what ```ReturnTypeWithAnnotations.MergeEquivalentTypes``` would do for custom modifiers

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +95,244 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                        hasParamChanges = true;
+                    }
+
+                    var mergedCustomModifiers = mergeCustomModifiers(thisParam.RefCustomModifiers, otherParam.RefCustomModifiers, paramVariance);
```
>var mergedCustomModifiers = mergeCustomModifiers(thisParam.RefCustomModifiers, otherParam.RefCustomModifiers, paramVariance); [](start = 20, length = 125)

Same comment as above about attempting to merge custom modifiers.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +95,244 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+
+            static ImmutableArray<CustomModifier> mergeCustomModifiers(ImmutableArray<CustomModifier> thisRefCustomModifiers, ImmutableArray<CustomModifier> otherRefCustomModifiers, VarianceKind variance)
+            {
+                Debug.Assert((thisRefCustomModifiers.IsDefault && otherRefCustomModifiers.IsDefault) || thisRefCustomModifiers.Length == otherRefCustomModifiers.Length);
```
>thisRefCustomModifiers.Length == otherRefCustomModifiers.Length [](start = 104, length = 63)

This is an incorrect assumption, there is no requirement for custom modifiers to match like that.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +95,244 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                return this;
+            }
+
+            static ImmutableArray<CustomModifier> mergeCustomModifiers(ImmutableArray<CustomModifier> thisRefCustomModifiers, ImmutableArray<CustomModifier> otherRefCustomModifiers, VarianceKind variance)
```
>variance [](start = 195, length = 8)

Variance isn't applicable to custom modifiers, they do not part of data flow information

333fred replied:
Function will be removed.

---
In reply to: [406963755](https://github.com/dotnet/roslyn/pull/43041#discussion_r406963755) [](ancestors = 406963755)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +95,244 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+        public FunctionPointerMethodSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnotations, TypeWithAnnotations> transform)
+        {
+            var transformedReturn = transform(ReturnTypeWithAnnotations);
+            var transformedRefCustomMods = transformRefCustomModifiers(transform, RefCustomModifiers);
```
>var transformedRefCustomMods = transformRefCustomModifiers(transform, RefCustomModifiers); [](start = 12, length = 90)

I do not think ever transform custom modifiers, they are not part of the language and meaning of any nullability information in them is undefined. See, for example how ```TypeWithAnnotations.SetUnknownNullabilityForReferenceTypes``` is implemented.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +95,244 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                        hasParamChanges = true;
+                    }
+
+                    var transformedCustomMods = transformRefCustomModifiers(transform, param.RefCustomModifiers);
```
>var transformedCustomMods = transformRefCustomModifiers(transform, param.RefCustomModifiers); [](start = 20, length = 93)

Same here for custom modifier transform

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +95,244 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                    }
+                    else
+                    {
+                        return thisRefCustomModifiers;
```
>return thisRefCustomModifiers [](start = 24, length = 29)

Leaking mergedModifiersBuilder?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +95,244 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                    }
+                    else
+                    {
+                        return existingRefCustomModifiers;
```
>return existingRefCustomModifiers [](start = 24, length = 33)

Leaking transformedModifiersBuilder?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -95,6 +95,244 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+                substitutedParameterTypes,
+                paramRefCustomModifiers);
+
+        internal FunctionPointerMethodSymbol MergeEquivalentTypes(FunctionPointerMethodSymbol signature, VarianceKind variance)
```
>MergeEquivalentTypes [](start = 45, length = 20)

Do we have tests that make the fixes around ref-ness observable?

333fred replied:
I'll add an extension to do this for ref as well, I have one for nonref

---
In reply to: [406970678](https://github.com/dotnet/roslyn/pull/43041#discussion_r406970678) [](ancestors = 406970678)

333fred replied:
Added `BestTypeForConditional_NestedParameterRef`

---
In reply to: [406983797](https://github.com/dotnet/roslyn/pull/43041#discussion_r406983797) [](ancestors = 406983797,406970678)

AlekseyTs commented:
Done with review pass (iteration 4)

333fred commented:
@AlekseyTs resolved feedback.

---
In reply to: [612260454](https://github.com/dotnet/roslyn/pull/43041#issuecomment-612260454) [](ancestors = 612260454)
